### ISSUE TITLE
Very high RAM Usage!

### ISSUE THREAD

**Issue by Ziya-Sharifli** [flybywiresim/simbridge#84]
SimBridge is using almost 7gb of ram on my 32gb PC. Is this normal? 
![simbridge issue](https://github.com/flybywiresim/simbridge/assets/129003936/f9800083-20ba-441c-a46b-a0002f9a9c75)

Latest development FBW A32NX, Latest SimBridge version.

**Comment by Seriox** [comment]
Got 32GB DDR5 RAM and Simbridge is using more than 15GB

![image](https://github.com/flybywiresim/simbridge/assets/6906690/27a0ecf8-2822-43c1-918a-7bd7456a37fe)

**Comment by CoffeePanda0** [comment]
Same here - seems to be a memory leak. RAM use increased from 100mb to 9GB after 2hrs before I killed it. Hopefully gets fixed soon as it's RAM use caused my sim to OOM.

**Comment by laszlo462** [comment]
Also seeing this issue.  After about an hour in flight, RAM usage is up to around 1500MB.  While this is less than the other high values above, by this point my 99% frames has dropped into the 10s.  Quite a jarring drop in performance for a 7800X3D with 32GB RAM and a 4080.

**Comment by Lucky38i** [comment]
The ND has been greatly refactored in the latest release. Can you please test and let me know?

**Comment by Lucky38i** [comment]
> Test, still goes up.
> 
> ![image](https://user-images.githubusercontent.com/4221827/265789536-50c640d7-fc5e-4325-a946-112fe141817c.png)
> 
> It was 57GB, when turned of terrain, after few mins, went to 40GB.

Committed RAM is not the same as actual RAM usage. What is the actual RAM usage

**Comment by deiks** [comment]
Just received CTD since didn't monitor memory. It increases "In use" memory and also commited. So def., issue is Terrain radar. Overall RAM usage was nearly full. When I got CTD, it wen't to 5-6 GB in idle.

**Comment by deiks** [comment]
From events:

`Windows successfully diagnosed a low virtual memory condition. The following programs consumed the most virtual memory: FlightSimulator.exe (6728) consumed 45770653696 bytes, fbw-simbridge.exe (4156) consumed 3046891520 bytes, and Navigraph Charts.exe (4816) consumed 1790185472 bytes.`

`Application popup: Windows - Out of Virtual Memory : Your system is low on virtual memory. To ensure that Windows runs properly, increase the size of your virtual memory paging file. For more information, see Help. `

**Comment by deiks** [comment]
Still active issue. Related to some specific areas.

Flew route `OEJN MIGDA2V MIGDA L677 CVO DCT HECA`

**Comment by BJ1960** [comment]
Still active this issue. Flight over Spain, FBW-Simbridge start with 2 GB RAM ussage and after 30 minutes flight now at 15 GB and ascending with every minute +1 GB more.

**Comment by deiks** [comment]
I downloaded the latest version and started local server. Received CTD with this error. Don't know if anything is helpful.

```
Error: vips_tracked: out of memory --- size == 1MB
vips2png: unable to write to target target
```

Before this, it was regular 

```[FBW-SimBridge] info    12/27/2023, 9:19:13 PM [McduGateway] Client connected +0ms```

**Comment by vajonam** [comment]
Had this today with 17G usage on the current version.

```CYYZ/05 N0441F380 BETES2 FOXEE DCT HVQ J85 SPA/N0438F390 DCT QUIWE/N0438F380 Q93 FISHO/N0441F390 Q93 GURGE/N0440F380 Q93 GIPPL Q85 LPERD SNFLD1 KMCO/17L```

**Comment by vajonam** [comment]
I wrote a powershell script to relaunch it after 18G of ram, and there is less than 5G for RAM free, this seems to be working well for me. 

```
$procName = "fbw-simbridge"
$RAMLimit = 18GB
$FREELimit = 5GB


while ($true) {
    $proc = Get-Process $procName -ea 0
    if ($proc -ne $null) {
         if ($proc.WorkingSet64 -gt $RAMLimit) {
        $free = get-ciminstance Win32_OperatingSystem | % FreePhysicalMemory
            if ($free*1024 -lt $FREELimit) {
           Write-Host "killing fbw-simbridge larger than $RAMLimit and free mem less than $free*1024"
               Stop-Process $proc -Force
               Start-Process E:\FlightSim2020\Community\flybywire-externaltools-simbridge\fbw-simbridge.exe  -ArgumentList "-hide"
            }
    }
    }
    Start-Sleep -s 30
}
```

**Comment by zaregg** [comment]
I have been having problems with CTDs and when looking at the Event Viewer I found this

![Image](https://github.com/user-attachments/assets/e08f3fa8-aeec-4499-8949-74458d4f1ad0)

Apperently i am running out of memory and Simbridge is using 18930782208 bytes aka ~18GB. That seems excessive

**Comment by deiks** [comment]
@zaregg works fine for me since the update a few months ago. It was something "rerendering map and feeding memory".
So hope you have your simbridge up to date.

**Comment by zaregg** [comment]
@deiks Yea i have the latest update of simbridge and recently deleted my whole sim because of the constant CTDs. Which I know found out are being caused by too little free virtual memory, and that simbridge uses like 20gb of virtual memory

**Comment by vajonam** [comment]
I find its related to enabling Terrain on ND, if you turn that off it lasts a while, but still have a script the kills it and relaunches it when grows to big.  attached above.

**Comment by deiks** [comment]
Yes, if you see your memory goes to high, just turn off Terrain and wait it to calm down. So you don't have to kill process.

Btw. Did you try with clean install of simbridge?

**Comment by zaregg** [comment]
> Yes, if you see your memory goes to high, just turn off Terrain and wait it to calm down. So you don't have to kill process.
> 
> Btw. Did you try with clean install of simbridge?

Yea I cleaned installed my whole sim and the flybywire installer

**Comment by frayien** [comment]
Hi ! The issue is still present, flybywire/simbridge only :

<img width="969" height="189" alt="Image" src="https://github.com/user-attachments/assets/2521b548-b715-44d3-9767-29876738bd25" />


With MSFS2020 running, in the main menu no flight started.

<img width="647" height="185" alt="Image" src="https://github.com/user-attachments/assets/16ded066-4caa-46af-b767-d6bb491a1b44" />

CPU : AMD Ryzen 9 7900X3D 12-Core Processor 4.40 GHz
RAM : 32,0 Go DDR5
SSD Samsung SSD 980 1TB
GPU : Radeon RX Vega (8 GB)
Windows 10 64 bits, x64

**Comment by frayien** [comment]
I have been investigating, it definitely comes from terrain, when I disable the TerrainModule the RAM usage drop to insignificance.
I will further investigate tomorrow, stay tuned !

**Comment by deiks** [comment]
> I have been investigating, it definitely comes from terrain, when I disable the TerrainModule the RAM usage drop to insignificance. I will further investigate tomorrow, stay tuned !

This issue is resolved while ago. I suggest to try clean install on simbridge.
Also it's not big issue if RAM goes up lil bit, the problem is when it crash the sim.

**Comment by frayien** [comment]
I'm using the latest possible version, freshly built from main. I tried to clean install before coming here anyway. I don't think 18+ Go to be insignificant or not an hindrance to the sim ...

**Comment by deiks** [comment]
> I'm using the latest possible version, freshly built from main. I tried to clean install before coming here anyway. I don't think 18+ Go to be insignificant or not an hindrance to the sim ...

In your case it's 9.2 Gb, but I agree it is a bit much. Takes more RAM then whole sim.

FlyByWire Installer (7) represents summary of processes below, so Simbridge takes 9 Gb.

**Comment by frayien** [comment]
Yeah I did not wait for it to fill up the RAM to take a screenshot :)
Honestly I am not sure my issue is the same as previously discussed here, it does not correlate to actually enabling terrain rendering in the sim.
Very weird stuff going on.

**Comment by frayien** [comment]
https://github.com/flybywiresim/simbridge/blob/main/apps/server/src/terrain/processing/maphandler.ts
line 301

On `updateGroundTruthPositionAndCachedTiles()`, `this.cachedElevationData.gpuData` is recreated but not deleted beforehand, 2Go memory leak every few minutes.

Also there seem to be two MapHandler instances alive do not know why.

Will continue to investigate why a function named `uploadWorldMapToGPU` takes 2Go of CPU memory, but this seem to be the most critical issue.

**Comment by frayien** [comment]
Something very wrong is going on, every call to `this.uploadWorldMapToGPU = this.uploadWorldMapToGPU.setOutput([worldWidth, worldHeight]);` is eating more and more RAM. Maybe something with gpu.js itself, maybe something related to GPU drivers

**Comment by tracernz** [comment]
https://github.com/gpujs/gpu.js/issues/841

Maybe better to just destroy the kernel and create a new one (without dynamic output) when the output size changes.

**Comment by Saschl** [comment]
> Something very wrong is going on, every call to `this.uploadWorldMapToGPU = this.uploadWorldMapToGPU.setOutput([worldWidth, worldHeight]);` is eating more and more RAM. Maybe something with gpu.js itself, maybe something related to GPU drivers

Nice find, may I ask which GPU and drivers you are using?

**Comment by frayien** [comment]
> [gpujs/gpu.js#841](https://github.com/gpujs/gpu.js/issues/841)
> 
> Maybe better to just destroy the kernel and create a new one (without dynamic output) when the output size changes.

Seems very similar indeed !

Note that I missunderstood some stuff, there is not multiple instances of MapHandler but the first time `updateGroundTruthPositionAndCachedTiles` is called the gpuCache was "cleared" beforehand (confusing me).

Note that calling `gpuCache.delete()` did not release any memory (at least in CPU RAM). I did not try to recreate the kernel.

I dont have my exact driver version at hand, my GPU is `Radeon RX Vega (8 GB)` (Should be the latest adrenaline version available)

Someone on Discord with a similar issue **it is not confirmed if his issue is the same** has a `MSI RTX 4060 8GB` with drivers `591.74 (studio drivers)`

**Comment by Saschl** [comment]
> > [gpujs/gpu.js#841](https://github.com/gpujs/gpu.js/issues/841)
> > Maybe better to just destroy the kernel and create a new one (without dynamic output) when the output size changes.
> 
> Seems very similar indeed !
> 
> Note that I missunderstood some stuff, there is not multiple instances of MapHandler but the first time `updateGroundTruthPositionAndCachedTiles` is called the gpuCache was "cleared" beforehand (confusing me).
> 
> Note that calling `gpuCache.delete()` did not release any memory (at least in CPU RAM). I did not try to recreate the kernel.
> 
> I dont have my exact driver version at hand, my GPU is `Radeon RX Vega (8 GB)` (Should be the latest adrenaline version available)
> 
> Someone on Discord with a similar issue **it is not confirmed if his issue is the same** has a `MSI RTX 4060 8GB` with drivers `591.74 (studio drivers)`

Feel free to try this changes, maybe it helps (I cannot really reproduce it reliably on my end, but maybe you can)

https://github.com/flybywiresim/simbridge/pull/145

**Comment by frayien** [comment]
Hummm... seems like we are filling the GPU RAM too
<img width="589" height="175" alt="Image" src="https://github.com/user-attachments/assets/fb7bfd1e-b7c9-465f-aa48-fab684f4dd33" />
On termination :
<img width="444" height="188" alt="Image" src="https://github.com/user-attachments/assets/6630b7a2-dd28-4408-adbe-dc427fff0426" />

You missed the calling `gpuCache.delete()` part before recreating the texture in your PR, but I added a `if (this.cachedElevationData.gpuData !== null) this.cachedElevationData.gpuData.delete();` before doing the above test.

Your fix seem to mitigate the `setOutput()` issue. But the call to 
```ts
this.cachedElevationData.gpuData = this.uploadWorldMapToGPU(
        this.cachedElevationData.cpuData,
        worldWidth,
      ) as Texture;
```
is still allocating gigabytes of RAM and never releasing anything. aka we fixed the least problematic of our issues...

Also my driver :
<img width="382" height="97" alt="Image" src="https://github.com/user-attachments/assets/9b4973f1-3ec8-4a3e-8198-12b641729849" />

Update : 
Checked the call to `gpuCache.delete()` and it seems the Texture is already deleted at this point. If I understand the documentation correctly calling `delete()` is only required when the `immutable` parameter is true, which it is not.
So I don't think the `gpuCache.delete()` issue is an issue at all.

Update :
Something very wrong is going on within gpu.js's Kernel ....

**Comment by SmallDragonAir** [comment]
I can reproduce what appears to be the same issue with the A32NX on MSFS 2024.

During a flight, fbw-simbridge.exe memory/commit usage kept increasing over time.

Windows Resource-Exhaustion-Detector (Event ID 2004) captured the following:

- 10:36:07 — fbw-simbridge.exe: 3,621,498,880 bytes (~3.37 GiB)
- 11:29:28 — fbw-simbridge.exe: 6,391,136,256 bytes (~5.95 GiB)

So SimBridge increased by about 2.6 GiB in ~53 minutes and did not appear to release the memory.

At the same time FlightSimulator2024.exe was using around 37–38 GiB of committed memory. The system eventually ran into virtual-memory exhaustion and later hard-froze/black-screened, requiring a forced shutdown.

System:
- MSFS 2024
- FlyByWire A32NX
- 16 GB RAM
- 48 GB page file
- RTX 5090 Laptop GPU
- Windows 11

I understand that the low physical RAM makes the final exhaustion easier to reach, but the continuously increasing SimBridge usage itself appears abnormal.

I can provide additional logs or collect a memory/handle usage trace if useful.

<img width="2560" height="1074" alt="Image" src="https://github.com/user-attachments/assets/7e853653-30a1-4cd1-9080-18529da2bd3b" />
